library/src/main/java/com/google/android/exoplayer2/text/ssa/SSADecoder.java

```diff
@@ -0,0 +1,235 @@
+    and with different ways of conveying timing information. Roughly speaking, an input buffer
+    received by a subtitle decoder consists of a timestamp (timeUs) and the subtitle data to be
+    decoded (data). There are four cases that can occur:
+
+    1. data contains all of the cues for the media and also their presentation timestamps.
+    timeUs is the time of the start of the media. The subtitle decoder receives a single input buffer.
+
+    2. data contains a single cue to be displayed at timeUs. There are no timestamps encoded in data.
+    The subtitle decoder receives many input buffers.
+
+    3. data contains cues covering a region of time (e.g. 5 seconds) along with their presentation
+    timestamps relative to the start of the region. timeUs is the time of the start of the region.
+    The subtitle decoder receives many input buffers.
+
+    4. As above, but the timestamps embedded in data are relative to the start of the media rather
+    than the start of the region. This case is tricky and best avoided.
+
+    For a side-loaded SSA file you'd have case (1).
+
+    For SSA embedded in MKV, it looks like they way it's embedded means you'd have case (2)
+    if you were to just pass the sample data through without changing it.
+    Note that timeUs is being set to blockTimeUs already.
+    Each region happens to be the duration of a single cue.
+
+    In the extractor, It's much easier to handle if you change the sample data so that you get case (3).
+    This basically means the embedded time should be 0 rather than blockTimeUs.
+
+    If you look at the SubRip case in the MKV extractor you'll see that it does exactly this.
+    The SubRip case also defers writing so that the end time can be set properly.
+
+    In the decoder you should create a new Subtitle instance for each decode call, rather than appending to an existing instance.
+
+    For the SSA embedded in MKV case you should end up with each call to decode producing a new Subtitle with a single cue at time 0.
+    The reason this works is that the event timing in a Subtitle is relative to timeUs of the buffer,
+    which is being set to blockTimeUs. When the decoder receives a new input buffer with a larger timeUs
+    than the previous one, the value passed to getCues will go down.
+ */
+
+
+public class SSADecoder extends SimpleSubtitleDecoder {
+    private static final String TAG = "SSADecoder";
+    private static String defaultDialogueFormat = "Start, End, Layer, Style, Name, MarginL, MarginR, MarginV, Effect, Text";
+    private static String defaultStyleFormat = "Name, Fontname, Fontsize, PrimaryColour, SecondaryColour, OutlineColour, BackColour, Bold, Italic, Underline, StrikeOut, ScaleX, ScaleY, Spacing, Angle, BorderStyle, Outline, Shadow, Alignment, MarginL, MarginR, MarginV, Encoding";
+    private String[] dialogueFormat;
+    private String[] styleFormat;
+    private Map<String,Style> styles = new HashMap<>();
```

Comment:
See other comment; you should probably put the header into the Format init data instead, if that works. This will require some plumbing of the format data from TextRenderer to the decoder.